test/concurrent/WaitLockTest.ooc

```diff
@@ -0,0 +1,113 @@
+/* This file is part of magic-sdk, an sdk for the open source programming language magic.
+ *
+ * Copyright (C) 2016 magic-lang
+ *
+ * This software may be modified and distributed under the terms
+ * of the MIT license.  See the LICENSE file for details.
+ */
+
+use concurrent
+use unit
+
+WaitLockTest: class extends Fixture {
+	init: func {
+		super("WithLock")
+		this add("_testWithMutexOwnership", This _testWithMutexOwnership)
+		this add("_testWithoutMutexOwnership", This _testWithoutMutexOwnership)
+		this add("_testWakeWithFailingCondition", This _testWakeWithFailingCondition)
+		this add("_testWakeWithPassingCondition", This _testWakeWithPassingCondition)
+	}
+	_testWithMutexOwnership: static func {
+		waitLock := WaitLock new()
+		waitLock free()
+	}
+	_testWithoutMutexOwnership: static func {
+		mutex := Mutex new()
+		waitLock := WaitLock new(mutex)
+		waitLock free()
+		mutex free()
+	}
+	_testWakeWithFailingCondition: static func {
+		timesTriggered := 0
+		timesTriggeredRef := timesTriggered&
+		waitLock := WaitLock new()
+		waitingThread := Thread new(||
+			waitLock lockWhen(||
+				timesTriggeredRef@ = timesTriggeredRef@ + 1
```

Comment:
It has to do with the closure variables not being passed by reference (could be due to it being a local variable?), therefore if the variable timesTriggered is updated the closure will not know about it.